tests/test_faster_rcnn.py

```diff
@@ -0,0 +1,377 @@
+import argparse
+import os
+from unittest.mock import MagicMock, patch
+
+import numpy as np
+import pytest
+import torch
+import torchvision
+from PIL import Image
+from tensorboardX import SummaryWriter
+from yacs.config import CfgNode as CN
+
+from datasetinsights.data.dummy.dummy_object_detection import DummyDetection2D
+from datasetinsights.estimators import Estimator
+from datasetinsights.estimators.faster_rcnn import (
+    collate_fn,
+    create_dataloader,
+    create_dryrun_dataset,
+    create_optimizer_lrs,
+    create_sampler,
+    download_data,
+    get_transform,
+    load_data,
+)
+from datasetinsights.storage.checkpoint import create_checkpointer
+
+TRAIN = "train"
+VAL = "val"
+TEST = "test"
+
+label_mappings = {"1": "car", "2": "bike"}
+dummy_data = DummyDetection2D(
+    transform=get_transform(), label_mappings=label_mappings
+)
+
+
+@pytest.fixture
+def config():
+    parser = argparse.ArgumentParser(
+        description="Datasetinsights Modeling Interface"
+    )
+    parser.add_argument(
+        "--config", type=str, required=True, help="config file for this model",
+    )
+    cmd_args = parser.parse_args(
+        ["--config=tests/configs/faster_rcnn_groceries_real_test.yaml"]
+    )
+    cfg = CN.load_cfg(open(cmd_args.config, "r"))
+
+    return cfg
+
+
+@patch("datasetinsights.estimators.faster_rcnn.Dataset.create")
+def test_faster_rcnn_train_one_epoch(mock_create, config):
+    mock_create.return_value = dummy_data
+    config.system.logdir = "/tmp/train_one/"
+    kfp_writer = MagicMock()
+    writer = SummaryWriter(config.system.logdir, write_to_disk=True)
+    checkpointer = create_checkpointer(logdir=writer.logdir, config=config)
+    estimator = Estimator.create(
+        config.estimator,
+        config=config,
+        writer=writer,
+        device=torch.device("cpu"),
+        checkpointer=checkpointer,
+        kfp_writer=kfp_writer,
+    )
+
+    train_dataset = download_data(config, TRAIN)
+    is_distributed = config.system.distributed
+    train_sampler = create_sampler(
+        is_distributed=is_distributed, dataset=train_dataset, is_train=True
+    )
+    train_loader = create_dataloader(
+        config, train_dataset, train_sampler, TRAIN
+    )
+
+    optimizer, lr_scheduler = create_optimizer_lrs(config, estimator.model)
+    accumulation_steps = config.train.get("accumulation_steps", 1)
+    estimator.train_one_epoch(
+        optimizer=optimizer,
+        data_loader=train_loader,
+        epoch=0,
+        lr_scheduler=lr_scheduler,
+        accumulation_steps=accumulation_steps,
+    )
+
+    assert os.listdir("/tmp/train_one/")[0].startswith("events.out.tfevents")
```

Comment:
Rely on output files that was specific for a training procedure is a test that need to be changed if the output format is somehow changed (for example,tensorboard api change). A better test usually requires validate the code execution with the mocked objects. 

For example, in this test, we can improve this test with some mocks:

```
def test_faster_rcnn_train_one_epoch(config):
    writer = Mock()
    kfp_writer = Mock()
    checkpointer = Mock()
    device = torch.device("cpu")

    estimator = FasterRCNN(
        config=config,
        writer=writer,
        device=device,
        checkpointer=checkpointer,
        kfp_writer=kfp_writer,
    )

    dataset = DummyDetection2D()
    sampler = torch.utils.data.RandomSampler(dataset)
    data_loader = torch.utils.data.DataLoader(dataset, sampler)

    # ideally this should be mocked as well...
    optimizer, lr_scheduler = create_optimizer_lrs(config, estimator.model)
    accumulation_steps = config.train.get("accumulation_steps", 1)
    epoch = 0
    estimator.train_one_epoch(
        optimizer=optimizer,
        data_loader=data_loader,
        epoch=epoch,
        lr_scheduler=lr_scheduler,
        accumulation_steps=accumulation_steps,
    )
    loss_value = 0.1
    with @patch("datasetinsights.estimator.faster_rcnn.Loss.compute") as mock_loss_compute:
        mock_loss_compute.return_value = loss_value
        writer.assert_called_with("training/loss", loss_value, epoch)
        writer.assert_called_with("training/running_loss", loss_value, epoch)

```

This is not a perfect solution either, as it might need to be fixed when the training code is changed. However, at least we know the test will not fail when external packages are changed. 

Does this make sense? Does this thinking apply to other test cases?

Reply:
I tried this, approach, by mocking both writer and checkpointer but below assertions didn't get called.    

writer.assert_called_with("training/loss", loss_value, epoch)
 writer.assert_called_with("training/running_loss", loss_value, epoch).